logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinExistingMLSConversationsUseCase.kt

```diff
@@ -0,0 +1,70 @@
+package com.wire.kalium.logic.feature.conversation
+
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.NetworkFailure
+import com.wire.kalium.logic.data.conversation.Conversation
+import com.wire.kalium.logic.data.conversation.ConversationRepository
+import com.wire.kalium.logic.data.conversation.Conversation.ProtocolInfo.MLS.GroupState
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.flatMapLeft
+import com.wire.kalium.logic.functional.foldToEitherWhileRight
+import com.wire.kalium.logic.kaliumLogger
+import com.wire.kalium.network.exceptions.KaliumException
+import com.wire.kalium.network.exceptions.isMlsStaleMessage
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.async
+
+/**
+ * Send an external add proposal to join all MLS conversations which the user is member
+ * of but has not yet joined the corresponding MLS group.
+ */
+class JoinExistingMLSConversationsUseCase(
+    val conversationRepository: ConversationRepository,
+    kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl
+) {
+    /**
+     * A dispatcher with limited parallelism of 10.
+     * This means using this dispatcher only a single coroutine will be processed at a time.
+     */
+    @Suppress("MagicNumber")
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private val dispatcher = kaliumDispatcher.default.limitedParallelism(10)
```

Comment:
@MohamadJaara would this also work as an alternative to the thread pool API?

Reply:
Not sure what was the need, but just my 2 cents here.

The dispatchers we use:
----

- `Default`: Targeted for general computation. By default is backed by 1 thread for each CPU core (but minimum = 2).
- `Main`: Currently only available on JVM (other platforms use `Default` underneath). It's the main. It has only 1 thread in its pool: the one used to draw the UI.
- `IO`: Currently only available on JVM. Targeted for Input/Output (_e.g._ networking, filesystem stuff). Defaults to 64 Threads or CPU core count (whichever is largest). When using `limitedParalellism`, it can create MORE threads if needed.

In general, in Kalium, we can use `Default` for most stuff. During networking, for example, Ktor will switch to its own dispatchers to performing requests. Even thought it's not _ideal_, we are using `Default` as `IO` for `iOS` for example. We don't use the file system so much that this would be performance costly anyway.


Back to the point
----

I'm not sure why the number `10`. Why not more, or why not less? But using `kaliumDispatcher.default.limitedParallelism(10)` will assure that no more than 10 coroutines will run in parallel.

If the `Default` dispatcher is backed by only 8 threads (8 CPU cores), the `limitedParallelism(10)` will not change anything at all, as it was already limited to 8 in the first place.

Reply:
My thinking was that a user potentially needs to join hundreds of MLS groups on login so I want to perform the requests in parallel. Now I don't want to perform 100 HTTP requests in parallel but if you say that Ktor anyway dispatches its network calls on it own dispatcher with it's own internal limits, maybe it fine to remove `limitedParallelism(10)` completely.

Reply:
This is indeed limiting to only 10 requests (or less depending on number of CPU cores).
I think we can just use `kaliumDispatcher.io` and leave Ktor to do its magic underneath.

If we notice any issues, we can come back and customise the Dispatcher to better suit our needs.

Also, out of curiosity, are there plans to optimise this in the future? I see one of MLS's goals is reducing unnecessary encryptions and backend requests compared to Proteus, being more scalable.

Maybe grouping together all the requests or similar.

This is definitely a point it is currently failing, although it only once after registering the client, so might be a price we're willing to pay anyway.

Reply:
>Maybe grouping together all the requests or similar.

Backends response to this is that it's unnecessary if we use HTTP2 multiplexing